activesupport/lib/active_support/event_reporter.rb

```diff
@@ -0,0 +1,499 @@
+  #   Rails.event.notify("user_created", { id: 123 })
+  #
+  #   # Emits event:
+  #   #  {
+  #   #    name: "user_created",
+  #   #    payload: { id: 123 },
+  #   #    context: { request_id: "abcd123", user_agent: TestAgent" },
+  #   #    timestamp: 1738964843208679035,
+  #   #    source_location: { filepath: "path/to/file.rb", lineno: 123, label: "UserService#create" }
+  #   #  }
+  #
+  # Context is reset automatically before and after each request.
+  #
+  # A custom context store can be configured via `config.active_support.event_reporter_context_store`.
+  #
+  #     # config/application.rb
+  #     config.active_support.event_reporter_context_store = CustomContextStore
+  #
+  #     class CustomContextStore
+  #       class << self
+  #         def context
+  #           # Return the context.
+  #         end
+  #
+  #         def set_context(context_hash)
+  #           # Append context_hash to the existing context store.
+  #         end
+  #
+  #         def clear
+  #           # Delete the stored context.
+  #         end
+  #       end
+  #     end
+  #
+  class EventReporter
+    attr_reader :subscribers
+    attr_accessor :raise_on_error
+
+    class << self
+      attr_accessor :context_store # :nodoc:
+
+      # Lookup an encoder by name or symbol
+      #
+      #   ActiveSupport::EventReporter.encoder(:json)
+      #   # => ActiveSupport::EventReporter::Encoders::JSON
+      #
+      #   ActiveSupport::EventReporter.encoder("msgpack")
+      #   # => ActiveSupport::EventReporter::Encoders::MessagePack
+      #
+      # ==== Arguments
+      #
+      # * +format+ - The encoder format as a symbol or string
+      #
+      # ==== Raises
+      #
+      # * +KeyError+ - If the encoder format is not found
+      def encoder(format)
+        encoders = {
+          json: Encoders::JSON,
+          msgpack: Encoders::MessagePack
+        }
+        encoders.fetch(format.to_sym) do
+          raise KeyError, "Unknown encoder format: #{format.inspect}. Available formats: #{encoders.keys.join(', ')}"
+        end
+      end
+    end
+
+    self.context_store = EventContext
+
+    def initialize(*subscribers, raise_on_error: false, tags: nil)
+      @subscribers = []
+      subscribers.each { |subscriber| subscribe(subscriber) }
+      @raise_on_error = raise_on_error
+    end
+
+    # Registers a new event subscriber. The subscriber must respond to
+    #
+    #   emit(event: Hash)
+    #
+    # The event hash will have the following keys:
+    #
+    #   name: String (The name of the event)
+    #   payload: Hash, Object (The payload of the event, or the event object itself)
+    #   tags: Hash (The tags of the event)
+    #   timestamp: Float (The timestamp of the event, in nanoseconds)
+    #   source_location: Hash (The source location of the event, containing the filepath, lineno, and label)
+    #
+    def subscribe(subscriber)
+      return if @subscribers.include?(subscriber)
```

Comment:
I wonder if it's surprising that trying to add a suscriber (that may already be added) be a no-op. How are they compared?